Firestore/core/test/unit/util/random_access_queue_test.cc

```diff
@@ -0,0 +1,647 @@
+/*
+ * Copyright 2017 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/util/random_access_queue.h"
+
+#include <iostream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+namespace {
+
+std::string ToDebugString(const std::vector<int>& elements) {
+  std::string result;
+  result += "[";
+  bool is_first_element = true;
+  for (int element : elements) {
+    if (is_first_element) {
+      is_first_element = false;
+    } else {
+      result += ", ";
+    }
+    result += std::to_string(element);
+  }
+  result += "]";
+  return result;
+}
+
+std::string ToDebugString(const RandomAccessQueue<int>& queue) {
+  return ToDebugString(queue.elements());
+}
+
+}  // namespace
+
+void PrintTo(const RandomAccessQueue<int>& queue, std::ostream* out) {
+  *out << ToDebugString(queue);
+}
+
+MATCHER_P(QueueElementsEqual,
+          expected_elements,
+          std::string("contains elements: ") +
+              ToDebugString(expected_elements)) {
+  return expected_elements == arg.elements();
```

Comment:
Can you use a built-in here? https://chromium.googlesource.com/external/gmock/+/master/test/gmock-generated-matchers_test.cc#387

Reply:
Done. Great suggestion.